lib/AST/Decl.cpp

```diff
@@ -4087,46 +4087,152 @@ bool ProtocolDecl::isAvailableInExistential(const ValueDecl *decl) const {
+          map[assocDecl] = true;
+      }
+    }
+  }
+}
+
+ExistentialSupportKind
+ProtocolDecl::existentialTypeSupportedSlow(LazyResolver *resolver,
+                                           AssociatedTypeFreedomMap &map,
+                                           bool needsCompleteAssocMap) {
+  if (Bits.ProtocolDecl.ExistentialTypeSupportedValid) {
+    auto supportKd = getExistentialSupportKind();
+    // This function recurses up the protocol hierarchy. If not
+    // supported due to free associated type member, we must let the
+    // inheriting protocol know about it.
+    if (supportKd == ExistentialSupportKind::UnsupportedAssoc ||
+        needsCompleteAssocMap)
+      getAssociatedTypeFreedomMapRec(map);
+    return supportKd;
+  }
+
+  auto setSupportKind = [&](ExistentialSupportKind ESK) {
+    Bits.ProtocolDecl.ExistentialTypeSupported = static_cast<unsigned>(ESK);
+  };
   // Assume for now that the existential type is supported; this
   // prevents circularity issues.
   Bits.ProtocolDecl.ExistentialTypeSupportedValid = true;
-  Bits.ProtocolDecl.ExistentialTypeSupported = true;
+  setSupportKind(ExistentialSupportKind::Supported);
 
-  // ObjC protocols can always be existential.
+  // ObjC protocols can always be existential and never have
+  // associated type members.
   if (isObjC())
-    return true;
+    return ExistentialSupportKind::Supported;
 
+  bool hasAssoc = false;
   for (auto member : getMembers()) {
-    // Check for associated types.
-    if (isa<AssociatedTypeDecl>(member)) {
-      // An existential type cannot be used if the protocol has an
-      // associated type.
-      Bits.ProtocolDecl.ExistentialTypeSupported = false;
-      return false;
+    if (auto assocDecl = dyn_cast<AssociatedTypeDecl>(member)) {
+      // Mark that we found an associated type and collect it.
+      hasAssoc = true;
+      map[assocDecl] = /*isConstrainedToConrete*/ false;
+      continue;
     }
 
-    // For value members, look at their type signatures.
+    // For value members, look for Self references in their type signatures.
     if (auto valueMember = dyn_cast<ValueDecl>(member)) {
       if (resolver && !valueMember->hasInterfaceType())
         resolver->resolveDeclSignature(valueMember);
 
-      if (!isAvailableInExistential(valueMember)) {
-        Bits.ProtocolDecl.ExistentialTypeSupported = false;
-        return false;
+      auto selfKind = findProtocolSelfReferences(
+                        valueMember, /*allowCovariantParameters=*/true,
+                        /*skipAssocTypes=*/true);
+
+      // A protocol with a Self reference in contravariant position
+      // cannot be used as an existential, alongside any inheriting
+      // protocol; compute the rest of the map if needed and bail out.
+      if (selfKind.parameter || selfKind.other) {
+        if (needsCompleteAssocMap)
+          getAssociatedTypeFreedomMapRec(map);
+        setSupportKind(ExistentialSupportKind::UnsupportedSelf);
```

Comment:
Set the support kind once in the caller, instead of setting it at every 'return' location here

Reply:
The function (`existentialTypeSupportedSlow`) is recursive, so if I set the support kind only in the first caller, protocols up the hierarchy will not have their support kinds set if the function happens to recurse. This wasn't a problem before because the function called `existentialTypeSupported` instead of directly recursing, but now `existentialTypeSupportedSlow` has a new map parameter. I don't think retaining an optional map parameter on `existentialTypeSupported` too just for the purpose of setting the support kind once is reasonable.